File: Heeseop An/2023-01-26-Longest-Repeating-Character-Replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        charCount = {} #frequency of each char
        result = 0
        l = 0 #left pointer

        for r in range(len(s)): #right pointer performing window slide
            charCount[s[r]] = 1 + charCount.get(s[r], 0) #increment char frequency and avoid default value error
            if (r - l + 1) - max(charCount.values()) > k:
                charCount[s[l]] -= 1
                l += 1
            result = max(result, (r - l + 1))

        return result

File: Dave Min/2023-01-26-LongestRepeatingCharacterReplacement.cs
public class Solution
{
    public int CharacterReplacement(string s, int k)
    {
        int[] frequencyMap = new int[26];
        int l = 0; int r = 0; //left: l,  right: r.
        int maxFrequency = 0;
        int longestSubstringLen = 0;

        while (r < s.Length)
        {
            int idx = (int)s[r] - 'A';
            int window = r - l + 1;
            frequencyMap[idx]++;
            maxFrequency = Math.Max(maxFrequency, frequencyMap[idx]);
            //if window - max > k, then l should be moved one to the right
            if (window - maxFrequency > k)
            {
                frequencyMap[(int)s[l] - 'A']--;

                l++;
            }
            longestSubstringLen = r + 1 - l;
            r++;
        }
        return longestSubstringLen;
    }
}